### PR TITLE
Nonzero simplex parameter control

### PR DESCRIPTION
**Pull request by AdrianSosic**
This PR adds the option to control the number of active (i.e. non-zero) parameters in the simplex construction of discrete search spaces.

### CODE REVIEW

**Comment by AdrianSosic** [comment]
Hi @Scienfitz. I have a whole group of sick people here at home, meaning that all rooms are blocked except my office 😆 🙈 Since there is nowhere else for me to go, I though I might just as well start drafting the simplex PR to get some work from your long TODO list. I think the current version should pretty much already do the job, but I haven't had time for super exhaustive testing. Hence, I'd really appreciate if you could give it a thorough review. Also, perhaps changing the nomenclature from "active" to just "nonzero" could be a little less ambiguous, especially since we already use "active" in another context? 

Anyway, hope this helps 👍🏼

**Comment by AdrianSosic** [comment]
Ah, and the variable naming and docstring of this part probably needs to be adjusted a little bit. Haven't had time to finish:
image
Originally only had the first variable but then noticed that the second is also needed. So that part isn't clean in terms of nomenclature and explanation. But functionally I think it should work

**Comment by AdrianSosic** [comment]
@Scienfitz, should be ready for merge now 🚀 But please still have a thorough look (also if the explanations etc make sense to you)

File: baybe/searchspace/discrete.py [hunk 1]
```diff
@@ -298,52 +308,113 @@ def from_simplex(
...
+                exp_rep,
+                max_sum=max_sum - min_sum_to_go,
+                # the maximum possible number of nonzeros to come dictates if we
+                # can achieve our minimum constraint in the end:
+                min_nonzero=min_nonzero - max_nonzero_to_go,
+                # the minimum possible number of nonzeros to come dictates if we
+                # can stay below the targeted maximum in the end:
+                max_nonzero=max_nonzero - min_nonzero_to_go,
+                boundary_only=False,
+            )
```
Comment (Scienfitz) [inline baybe/searchspace/discrete.py:417]: is there a bug down in line 430?

it says
`cls(parameters=simplex_parameters, exp_rep=exp_rep)`

but shouldnt it be 
`cls(parameters=simplex_parameters+product_parameters, exp_rep=exp_rep)`

**Review by Scienfitz — COMMENTED** [review]
thanks for doing this despite ooo, really cool functionality that we might also immediately need for some projects

File: baybe/searchspace/discrete.py (hunk 1 above)
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:417]: (if im right and this is a bug the tests dont capture this at the moment)

File: baybe/searchspace/discrete.py [hunk 2]
```diff
@@ -298,52 +308,113 @@ def from_simplex(
 
         # Validate non-negativity
         min_values = [min(p.values) for p in simplex_parameters]
+        max_values = [max(p.values) for p in simplex_parameters]
```
Comment (AVHopp) [inline baybe/searchspace/discrete.py:311]: The comment for this part 2 lines above says `# Validate non-negativity`, this should be adjusted.

File: baybe/searchspace/discrete.py [hunk 3]
```diff
@@ -298,52 +308,113 @@ def from_simplex(
...
+            max_nonzero: Optional[int] = None,
+        ) -> None:
+            """Drop rows that violate the specified simplex constraint.
 
             Args:
                 df: The dataframe whose rows should satisfy the simplex constraint.
                 max_sum: The maximum row sum defining the simplex size.
                 boundary_only: Flag to control if the points represented by the rows
                     may lie inside the simplex or on its boundary only.
+                min_nonzero: Minimum number of nonzero parameters allowed per row.
```
Comment (AVHopp) [inline baybe/searchspace/discrete.py:332]: Isn't this rather `Minimum number of nonzero parameters required per row`? Or do I understand this wrong?

File: baybe/searchspace/discrete.py [hunk 4]
```diff
@@ -298,52 +308,113 @@ def from_simplex(
...
+                max_nonzero: Maximum number of nonzero parameters allowed per row.
             """
+            # Apply sum constraints
             row_sums = df.sum(axis=1)
             if boundary_only:
-                locs_to_drop = row_sums[
-                    (row_sums < max_sum - tolerance) | (row_sums > max_sum + tolerance)
-                ].index
+                violated = (row_sums < max_sum - tolerance) | (
```
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/searchspace/discrete.py:338]: I'd also appreciate this name.

File: baybe/searchspace/discrete.py [hunk 5]
```diff
@@ -298,52 +308,113 @@ def from_simplex(
...
+                )
             else:
-                locs_to_drop = row_sums[row_sums > max_sum + tolerance].index
+                violated = row_sums > max_sum + tolerance
+
+            # Apply optional nonzero constraints
+            if (min_nonzero is not None) or (max_nonzero is not None):
+                n_nonzero = (df != 0.0).sum(axis=1)
+                if min_nonzero is not None:
+                    min_nonzero_violated = n_nonzero < min_nonzero
```
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/searchspace/discrete.py:348]: Would be fine for me, no strong opinion.

File: baybe/searchspace/discrete.py [hunk 6]
```diff
@@ -298,52 +308,113 @@ def from_simplex(
...
             if boundary_only:
-                locs_to_drop = row_sums[
-                    (row_sums < max_sum - tolerance) | (row_sums > max_sum + tolerance)
-                ].index
+                violated = (row_sums < max_sum - tolerance) | (
+                    row_sums > max_sum + tolerance
+                )
             else:
-                locs_to_drop = row_sums[row_sums > max_sum + tolerance].index
+                violated = row_sums > max_sum + tolerance
```
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/searchspace/discrete.py:342]: Would be fine for me, no strong opinion.

File: baybe/searchspace/discrete.py [hunk 7]
```diff
@@ -298,52 +308,113 @@ def from_simplex(
...
+        # * Similarly, it can be verified for each parameter that there are still
+        #   enough nonzero parameters to come to even reach the minimum
+        #   desired number of nonzero after all joins.
+        for i, (
+            param,
+            min_sum_to_go,
+            min_nonzero_to_go,
+            max_nonzero_to_go,
+        ) in enumerate(
+            zip_longest(
```
Comment (AVHopp) [inline baybe/searchspace/discrete.py:393]: Why is the `zip_longest` necessary, shouldn't all of these have the same length?

Comment (AVHopp) [reply baybe/searchspace/discrete.py:393]: Or this due to the fact that there might be other parameters like Categorical ones that do not contribute here?

File: baybe/searchspace/discrete.py (hunk 1 above)
Comment (AVHopp) [reply baybe/searchspace/discrete.py:417]: From what I understand here, the `product_parameters` are included in the `exp_rep` (line 424) but not in the construction of the class. So I would agree that this is a bug here.

File: tests/hypothesis_strategies/alternative_creation/test_searchspace.py [hunk 8]
```diff
@@ -162,3 +162,21 @@ def test_discrete_space_creation_from_simplex_mixed(
...
+    """The number of nonzero simplex parameters is controllable."""
+    params = [
+        NumericalDiscreteParameter(str(i), np.linspace(0, 1, 11)) for i in range(10)
+    ]
+    subspace = SubspaceDiscrete.from_simplex(
+        max_sum=1.0,
+        simplex_parameters=params,
+        min_nonzero=2,
+        max_nonzero=4,
+        boundary_only=True,
```
[the remark this answers is not in the document]
Comment (AVHopp) [reply tests/hypothesis_strategies/alternative_creation/test_searchspace.py:177]: Agree that it should be parameterized to test both.

File: baybe/searchspace/discrete.py (hunk 1 above)
Comment (AVHopp) [reply baybe/searchspace/discrete.py:417]: It seems like there is no validation function for the exp_rep, right? I guess we should add some check that verifies that the columns of exp rep and the names of the parameters are identical as a set, right?

**Review by AVHopp — COMMENTED** [review]
Looks mostly good to me, there is only the bug that we need to discuss here

File: tests/hypothesis_strategies/alternative_creation/test_searchspace.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply tests/hypothesis_strategies/alternative_creation/test_searchspace.py:177]: changed

File: baybe/searchspace/discrete.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:342]: changed

File: baybe/searchspace/discrete.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:348]: changed

File: baybe/searchspace/discrete.py [hunk 9]
```diff
@@ -298,52 +308,113 @@ def from_simplex(
...
+                n_nonzero = (df != 0.0).sum(axis=1)
+                if min_nonzero is not None:
+                    min_nonzero_violated = n_nonzero < min_nonzero
+                    violated |= min_nonzero_violated
+                if max_nonzero is not None:
+                    max_nonzero_violated = n_nonzero > max_nonzero
+                    violated |= max_nonzero_violated
+
+            # Remove violating rows
+            locs_to_drop = df[violated].index
```
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:355]: changed

File: baybe/searchspace/discrete.py (hunk 3 above)
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:332]: i think its good to say the minimum is required and theres a maximum allowed so i changed it

File: baybe/searchspace/discrete.py (hunk 2 above)
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:311]: think it refers to what comes in the next line 312+

File: baybe/searchspace/discrete.py (hunk 7 above)
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:393]: hthink it has to do with cumsum yielding a length-1 of the pbject you pass to it, that is prob trvially 0 hence the fillup

**Review by Scienfitz — APPROVED** [review]

**Pull request merged**